### PR TITLE
343 move part

### PR DESCRIPTION
**Pull request by raduiordache**

### CODE REVIEW

File: pykechain/client.py [hunk 1]
```diff
@@ -1061,6 +1061,41 @@ def create_model(self, parent, name, multiplicity='ZERO_MANY', **kwargs):
 
         return self._create_part(action="create_child_model", data=data, **kwargs)
 
+    def _create_clone(self, parent, part, **kwargs):
+        """
```
Comment (jberends) [inline pykechain/client.py]: please change docs to conform to other method documentation.

Add an example

File: pykechain/extra_utils.py [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+from pykechain.utils import temp_chdir
+
+# global variable
+__mapping_dictionary = None
+__edited_one_many = list()
+
+
+def get_mapping_dictionary(clean=False):
+    """
+    Get a temprorary helper to map some keys to some values. Mainly used in the relocation of parts and models.
```
Comment (jberends) [inline pykechain/extra_utils.py]: spelling of `temprorary`

File: pykechain/extra_utils.py [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+        # Populate the descendants of the Part (category=Instance), in order to avoid to retrieve children for every
+        # level and save time. Only need it the children should be included.
+        if include_children:
+            part.populate_descendants()
+
+        # This function will move the part instance under the target_parent instance, and its children if required.
+        moved_instance = move_part_instance(part_instance=part, target_parent=target_parent, part_model=part_model,
+                                            name=name, include_children=include_children)
+        return moved_instance
+    else:
```
Comment (jberends) [inline pykechain/extra_utils.py]: move the exception check to a two line check in the top

File: pykechain/extra_utils.py [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+        elif prop_instance._json_data['property_type'] == PropertyType.REFERENCES_VALUE:
+            if prop_instance.value:
+                moved_prop_instance = get_mapping_dictionary()[prop_instance.id]
+                properties_id_dict[moved_prop_instance.id] = [ref_part.id for ref_part in prop_instance.value]
+        else:
+            moved_prop_instance = get_mapping_dictionary()[prop_instance.id]
+            properties_id_dict[moved_prop_instance.id] = prop_instance.value
+    # Update the name and property values in one go.
+    moved_instance.update(name=name, update_dict=properties_id_dict, bulk=True, suppress_kevents=True)
+    return
```
Comment (jberends) [inline pykechain/extra_utils.py]: empty return, please provide an object not do not return. document as well.

File: pykechain/models/part.py [hunk 5]
```diff
@@ -717,3 +718,154 @@ def populate_descendants(self, batch=200):
                     parent._cached_children.append(child)
 
         self._cached_children = descendants_flat_list[0]._cached_children
+
+    def clone(self, **kwargs):
```
Comment (jberends) [inline pykechain/models/part.py]: add documentation and example.

add `.. versionadded:: 2.3`

File: pykechain/models/part.py [hunk 6]
```diff
@@ -717,3 +718,154 @@ def populate_descendants(self, batch=200):
...
+        parent = self.parent()
+        return self._client._create_clone(parent, self, **kwargs)
+
+    def copy_model(self, target_parent, name=None, include_children=True, include_instances=True):
```
Comment (jberends) [inline pykechain/models/part.py]: can't we just use `copy()` and based if it an instance or model do the math?

File: pykechain/models/part.py [hunk 7]
```diff
@@ -717,3 +718,154 @@ def populate_descendants(self, batch=200):
...
+            parent_instances = list(target_parent.instances())
+            for parent_instance in parent_instances:
+                for instance in instances_to_be_copied:
+                    instance.populate_descendants()
+                    move_part_instance(part_instance=instance, target_parent=parent_instance,
+                                       part_model=self, name=instance.name, include_children=include_children)
+
+        return copied_model
+
+    def move_model(self, target_parent, name=None, include_children=True, include_instances=True):
```
Comment (jberends) [inline pykechain/models/part.py]: can't we just use `move()` and if it is an instance or model do the math in the helpers?

File: pykechain/models/part.py [hunk 8]
```diff
@@ -717,3 +718,154 @@ def populate_descendants(self, batch=200):
...
+            retrieve_parent_instances = list(target_parent.instances())
+            for parent_instance in retrieve_parent_instances:
+                for instance in retrieve_instances_to_copied:
+                    instance.populate_descendants()
+                    move_part_instance(part_instance=instance, target_parent=parent_instance,
+                                       part_model=self, name=instance.name, include_children=include_children)
+        self.delete()
+        return moved_model
+
+    def copy_instance(self, target_parent, name=None, include_children=True):
```
Comment (jberends) [inline pykechain/models/part.py]: see remarks about `copy()`

File: pykechain/models/part.py [hunk 9]
```diff
@@ -717,3 +718,154 @@ def populate_descendants(self, batch=200):
...
+
+        """
+        #TODO: refactor out the illegality check of the copy in the helpers
+        if self.model().id not in [ target_parent.model().id, ].append([c.id for c in self.children()]):
+            raise IllegalArgumentError('Part and target_parent have the same model.')
+        copied_instance = relocate_instance(part=self, target_parent=target_parent, name=name,
+                                            include_children=include_children)
+        return copied_instance
+
+    def move_instance(self, target_parent, name=None, include_children=True):
```
Comment (jberends) [inline pykechain/models/part.py]: see other remarks about `mve()`

File: pykechain/utils.py [hunk 10]
```diff
@@ -28,3 +32,26 @@ def is_uuid(value):
         return True
     else:
         return False
+
+
+@contextmanager
+def temp_chdir(cwd=None):
```
Comment (jberends) [inline pykechain/utils.py]: we might need to add docs here as it is now inside the main pykechain world

Comment (jberends) [reply pykechain/utils.py]: resolved

**Comment by raduiordache** [comment]
@jberends can you re-review this now, please?

**Comment by raduiordache** [comment]
Just dumping a reminder for myself to take into consideration. Add 'multiplicity' argument that can be set by the user on the relocated `Part`. It should default to '`Part`.multiplicity' and should only be settable on the `Part` you are relocating, not its subparts. The idea comes from the discussion on [KECSUPPORT-114](https://support.ke-chain.com/jira/projects/KECSUPPORT/queues/custom/1/KECSUPPORT-114).

**Pull request merged**